File: typescript/design_pattern/composite/main_composite.ts
/**
 * Composite Design Pattern Implementation
 *
 * This example demonstrates the Composite pattern using a file system hierarchy.
 * Files are leaf nodes and Directories are composite nodes that can contain
 * both files and other directories.
 */

// Component interface - declares common operations for both leaf and composite objects
interface FileSystemComponent {
    getName(): string;
    getSize(): number;
    display(indent?: string): void;
    search(name: string): FileSystemComponent[];
}

// Leaf class - represents end objects that have no children
class FileNode implements FileSystemComponent {
    private name: string;
    private size: number;

    constructor(name: string, size: number) {
        this.name = name;
        this.size = size;
    }

    getName(): string {
        return this.name;
    }

    getSize(): number {
        return this.size;
    }

    display(indent: string = ""): void {
        console.log(`${indent}[Line 36] File: ${this.name} (${this.size} KB)`);
    }

    search(name: string): FileSystemComponent[] {
        if (this.name.toLowerCase().includes(name.toLowerCase())) {
            return [this];
        }
        return [];
    }
}

// Composite class - represents complex components that may have children
class Directory implements FileSystemComponent {
    private name: string;
    private children: FileSystemComponent[] = [];

    constructor(name: string) {
        this.name = name;
    }

    getName(): string {
        return this.name;
    }

    // Calculates total size by summing all children recursively
    getSize(): number {
        return this.children.reduce((total, child) => total + child.getSize(), 0);
    }

    add(component: FileSystemComponent): void {
        this.children.push(component);
    }

    remove(component: FileSystemComponent): void {
        const index = this.children.indexOf(component);
        if (index > -1) {
            this.children.splice(index, 1);
        }
    }

    getChildren(): FileSystemComponent[] {
        return this.children;
    }

    display(indent: string = ""): void {
        console.log(`${indent}[Line 81] Directory: ${this.name}/ (${this.getSize()} KB total)`);
        this.children.forEach(child => {
            child.display(indent + "  ");
        });
    }

    // Searches recursively through all children
    search(name: string): FileSystemComponent[] {
        let results: FileSystemComponent[] = [];

        if (this.name.toLowerCase().includes(name.toLowerCase())) {
            results.push(this);
        }

        this.children.forEach(child => {
            results = results.concat(child.search(name));
        });

        return results;
    }
}

// Client code that works with all components through the base interface
function printSeparator(title: string): void {
    console.log("\n" + "=".repeat(50));
    console.log(`[Line 106] ${title}`);
    console.log("=".repeat(50));
}

// Main demonstration
function main(): void {
    console.log("[Line 112] Creating file system structure...\n");

    // Create leaf nodes (files)
    const file1 = new FileNode("readme.md", 5);
    const file2 = new FileNode("index.ts", 12);
    const file3 = new FileNode("styles.css", 8);
    const file4 = new FileNode("app.ts", 25);
    const file5 = new FileNode("utils.ts", 15);
    const file6 = new FileNode("config.json", 3);
    const file7 = new FileNode("package.json", 2);
    const file8 = new FileNode("test.spec.ts", 18);

    // Create composite nodes (directories)
    const root = new Directory("project");
    const src = new Directory("src");
    const components = new Directory("components");
    const tests = new Directory("tests");
    const config = new Directory("config");

    // Build the tree structure
    console.log("[Line 132] Building directory tree...");

    // Add files to components directory
    components.add(file3);
    components.add(file4);

    // Add files and subdirectory to src
    src.add(file2);
    src.add(file5);
    src.add(components);

    // Add test files
    tests.add(file8);

    // Add config files
    config.add(file6);

    // Build root directory
    root.add(file1);
    root.add(file7);
    root.add(src);
    root.add(tests);
    root.add(config);

    // Display the complete tree structure
    printSeparator("Complete File System Structure");
    root.display();

    // Demonstrate uniform treatment - both files and directories respond to getSize()
    printSeparator("Size Calculations (Uniform Treatment)");
    console.log(`[Line 162] Root directory total size: ${root.getSize()} KB`);
    console.log(`[Line 163] src directory size: ${src.getSize()} KB`);
    console.log(`[Line 164] components directory size: ${components.getSize()} KB`);
    console.log(`[Line 165] Single file (app.ts) size: ${file4.getSize()} KB`);

    // Demonstrate search functionality
    printSeparator("Search Functionality");

    const tsFiles = root.search(".ts");
    console.log(`[Line 171] Found ${tsFiles.length} items containing '.ts':`);
    tsFiles.forEach(item => {
        console.log(`  - ${item.getName()} (${item.getSize()} KB)`);
    });

    const configSearch = root.search("config");
    console.log(`\n[Line 177] Found ${configSearch.length} items containing 'config':`);
    configSearch.forEach(item => {
        console.log(`  - ${item.getName()}`);
    });

    // Demonstrate adding and removing components
    printSeparator("Dynamic Modifications");

    const newFile = new FileNode("api.ts", 20);
    src.add(newFile);
    console.log(`[Line 187] Added new file 'api.ts' to src directory`);
    console.log(`[Line 188] Updated src directory size: ${src.getSize()} KB`);
    console.log(`[Line 189] Updated root directory size: ${root.getSize()} KB`);

    // Remove a file
    src.remove(newFile);
    console.log(`[Line 193] Removed 'api.ts' from src directory`);
    console.log(`[Line 194] Reverted src directory size: ${src.getSize()} KB`);

    // Demonstrate that client code doesn't need to know whether it's
    // working with a leaf or composite
    printSeparator("Polymorphic Behavior");

    const components_to_check: FileSystemComponent[] = [file1, src, components, file4];
    console.log("[Line 201] Checking sizes of mixed components (files and directories):");
    components_to_check.forEach(component => {
        console.log(`  ${component.getName()}: ${component.getSize()} KB`);
    });

    // Display nested structure for clarity
    printSeparator("Final Directory Structure");
    root.display();

    console.log("\n[Line 210] Composite pattern demonstration complete!");
}

// Run the demonstration
main();

File: rust/design_pattern/composite/main_composite.rs
// Composite Design Pattern in Rust
// Demonstrates a file system hierarchy with files (leaves) and directories (composites)

use std::cell::RefCell;
use std::rc::Rc;

// Component trait - defines the interface for all objects in the composition
trait FileSystemComponent {
    fn get_name(&self) -> &str;
    fn get_size(&self) -> u64;
    fn display(&self, indent: usize);
    fn is_composite(&self) -> bool {
        false
    }
}

// Leaf - represents individual files that cannot contain other components
struct File {
    name: String,
    size: u64,
}

impl File {
    fn new(name: &str, size: u64) -> Self {
        println!("  [Created File] '{}' ({} bytes)", name, size);
        File {
            name: name.to_string(),
            size,
        }
    }
}

impl FileSystemComponent for File {
    fn get_name(&self) -> &str {
        &self.name
    }

    fn get_size(&self) -> u64 {
        self.size
    }

    fn display(&self, indent: usize) {
        let prefix = "  ".repeat(indent);
        println!("{}|- File: {} ({} bytes)", prefix, self.name, self.size);
    }
}

// Composite - represents directories that can contain other components
struct Directory {
    name: String,
    children: RefCell<Vec<Rc<dyn FileSystemComponent>>>,
}

impl Directory {
    fn new(name: &str) -> Self {
        println!("  [Created Directory] '{}'", name);
        Directory {
            name: name.to_string(),
            children: RefCell::new(Vec::new()),
        }
    }

    fn add(&self, component: Rc<dyn FileSystemComponent>) {
        println!("  [Added] '{}' to '{}'", component.get_name(), self.name);
        self.children.borrow_mut().push(component);
    }

    fn remove(&self, name: &str) {
        self.children.borrow_mut().retain(|c| c.get_name() != name);
        println!("  [Removed] '{}' from '{}'", name, self.name);
    }
}

impl FileSystemComponent for Directory {
    fn get_name(&self) -> &str {
        &self.name
    }

    fn get_size(&self) -> u64 {
        // Recursively calculate total size of all children
        self.children
            .borrow()
            .iter()
            .map(|child| child.get_size())
            .sum()
    }

    fn display(&self, indent: usize) {
        let prefix = "  ".repeat(indent);
        println!("{}|+ Directory: {} (total: {} bytes)", prefix, self.name, self.get_size());

        // Recursively display all children
        for child in self.children.borrow().iter() {
            child.display(indent + 1);
        }
    }

    fn is_composite(&self) -> bool {
        true
    }
}

fn main() {
    println!("=== Composite Design Pattern Demo ===\n");

    // Step 1: Create the root directory
    println!("Step 1: Creating root directory structure");
    let root = Rc::new(Directory::new("root"));

    // Step 2: Create files for root directory
    println!("\nStep 2: Creating files for root");
    let readme = Rc::new(File::new("README.md", 1024));
    let config = Rc::new(File::new("config.json", 512));

    // Step 3: Add files to root
    println!("\nStep 3: Adding files to root");
    root.add(readme);
    root.add(config);

    // Step 4: Create subdirectories
    println!("\nStep 4: Creating subdirectories");
    let src_dir = Rc::new(Directory::new("src"));
    let tests_dir = Rc::new(Directory::new("tests"));

    // Step 5: Create files for src directory
    println!("\nStep 5: Creating files for src directory");
    let main_file = Rc::new(File::new("main.rs", 2048));
    let lib_file = Rc::new(File::new("lib.rs", 4096));
    let utils_file = Rc::new(File::new("utils.rs", 1536));

    // Step 6: Add files to src directory
    println!("\nStep 6: Adding files to src directory");
    src_dir.add(main_file);
    src_dir.add(lib_file);
    src_dir.add(utils_file);

    // Step 7: Create nested directory inside src
    println!("\nStep 7: Creating nested directory (modules)");
    let modules_dir = Rc::new(Directory::new("modules"));
    let auth_file = Rc::new(File::new("auth.rs", 3072));
    let db_file = Rc::new(File::new("database.rs", 2560));

    println!("\nStep 8: Adding files to modules directory");
    modules_dir.add(auth_file);
    modules_dir.add(db_file);

    // Step 9: Add modules directory to src
    println!("\nStep 9: Adding modules to src directory");
    src_dir.add(modules_dir);

    // Step 10: Create test files
    println!("\nStep 10: Creating test files");
    let test_main = Rc::new(File::new("test_main.rs", 1024));
    let test_utils = Rc::new(File::new("test_utils.rs", 768));

    // Step 11: Add test files to tests directory
    println!("\nStep 11: Adding test files to tests directory");
    tests_dir.add(test_main);
    tests_dir.add(test_utils);

    // Step 12: Add subdirectories to root
    println!("\nStep 12: Adding subdirectories to root");
    root.add(src_dir);
    root.add(tests_dir);

    // Step 13: Display the complete tree structure
    println!("\n=== Complete File System Tree ===\n");
    root.display(0);

    // Step 14: Demonstrate uniform treatment - get sizes
    println!("\n=== Size Calculations (Uniform Treatment) ===\n");
    println!("Total size of root: {} bytes", root.get_size());

    // Step 15: Create a standalone file to show leaf behavior
    println!("\n=== Demonstrating Leaf vs Composite ===\n");
    let standalone = Rc::new(File::new("standalone.txt", 256));
    println!("Standalone file size: {} bytes", standalone.get_size());
    println!("Is standalone a composite? {}", standalone.is_composite());
    println!("Is root a composite? {}", root.is_composite());

    // Step 16: Demonstrate remove operation
    println!("\n=== Demonstrating Remove Operation ===\n");
    let temp_dir = Rc::new(Directory::new("temp"));
    let temp_file = Rc::new(File::new("temp.txt", 100));
    temp_dir.add(temp_file);
    println!("Temp directory size before removal: {} bytes", temp_dir.get_size());
    temp_dir.remove("temp.txt");
    println!("Temp directory size after removal: {} bytes", temp_dir.get_size());

    println!("\n=== End of Composite Pattern Demo ===");
}
